include/multipass/ssh/sftp_client.h

```diff
@@ -18,31 +18,55 @@
 #ifndef MULTIPASS_SFTP_CLIENT_H
 #define MULTIPASS_SFTP_CLIENT_H
 
-#include <multipass/ssh/ssh_session.h>
+#include "ssh_session.h"
 
 #include <libssh/sftp.h>
 
+#include <filesystem>
+#include <functional>
 #include <iostream>
-#include <memory>
-#include <string>
+
+#include <QFlags>
+
+namespace fs = std::filesystem;
 
 namespace multipass
 {
 using SSHSessionUPtr = std::unique_ptr<SSHSession>;
-using SFTPSessionUPtr = std::unique_ptr<sftp_session_struct, void (*)(sftp_session)>;
+using SFTPSessionUPtr = std::unique_ptr<sftp_session_struct, std::function<void(sftp_session)>>;
+
+enum TransferFlags
+{
+    Recursive = 1,
+};
+
+SFTPSessionUPtr make_sftp_session(ssh_session session);
 
 class SFTPClient
 {
 public:
+    SFTPClient() = default;
     SFTPClient(const std::string& host, int port, const std::string& username, const std::string& priv_key_blob);
     SFTPClient(SSHSessionUPtr ssh_session);
 
-    void push_file(const std::string& source_path, const std::string& destination_path);
-    void pull_file(const std::string& source_path, const std::string& destination_path);
-    void stream_file(const std::string& destination_path, std::istream& cin);
-    void stream_file(const std::string& source_path, std::ostream& cout);
+    virtual bool is_dir(const fs::path& path);
+    virtual bool push(const fs::path& source_path, const fs::path& target_path, const QFlags<TransferFlags> flags,
+                      std::ostream& err_sink);
+    virtual bool pull(const fs::path& source_path, const fs::path& target_path, const QFlags<TransferFlags> flags,
```

Comment:
Okay, so here it was a complicated situation.

I think that when an error occurs, the user should be notified, but the transfer should still keep going and copy as many files as it can. That's why I think we can't let these functions throw exceptions when an error occurs, since that would interrupt the whole transfer.

Another alternative would be to accumulate all errors inside the function and throw a big exception at the end with all the errors. But then the errors wouldn't be reported in real time, only when the whole transfer is finished.

My approach was to use the `err_sink` to immediately be able to report errors as they occur, then return a success/failure bool at the end. This was the best way I could think of to accomplish a system that doesn't quit at the first error and tries to copy as much as it can, but that also reports all errors live, not just at the end.

Reply:
The point is that we only want to continue copying on expected errors, I believe. If there is an unexpected error, we should instead fail early and interrupt. Continuing normally on unexpected errors would make catching bugs that cascaded from them harder to understand and link to the initial issue. If, however, we find new error cases that we want to continue copying on, we can always start treating them as expected too, selectively. So, if this were critical software, I would say the "best effort" approach that you went for would be the best option. In our case, I think "fail early" is better.

As to how to achieve that, and aiming for simplicity (fewer parts to break, less code to understand, maintain...), I would avoid using Qt signals. It would be another mechanism that could fail when we were already handling a failure. And I think it would add unnecessary complexity. 

We have everything we need built in C++ itself. I tried to describe how in [other](https://github.com/canonical/multipass/pull/2639/#discussion_r959507521) [comments](https://github.com/canonical/multipass/pull/2639/#discussion_r959529597), probably not very clearly... The idea would be to throw dedicated exceptions (e.g. `SFTPException`) on problems that we can recover from and continue on. Then, the functions that now return `bool` would catch those dedicated exceptions, accumulate the relevant messages but otherwise continue with the copy, and then throw again when done (using `fmt::join`, to compose the final message, for example). Other exceptions would just be let through.